presto-main/src/main/java/com/facebook/presto/type/MapType.java

```diff
@@ -29,36 +29,50 @@
 import java.util.Map;
 
 import static com.facebook.presto.type.TypeUtils.appendToBlockBuilder;
-import static com.facebook.presto.type.TypeUtils.buildStructuralSlice;
 import static com.facebook.presto.type.TypeUtils.checkElementNotNull;
 import static com.facebook.presto.type.TypeUtils.hashPosition;
 import static com.facebook.presto.type.TypeUtils.parameterizedTypeName;
-import static com.facebook.presto.type.TypeUtils.readStructuralBlock;
 import static com.google.common.base.Preconditions.checkArgument;
 
 public class MapType
-        extends AbstractVariableWidthType
+        extends AbstractType
 {
     private final Type keyType;
     private final Type valueType;
     private static final String MAP_NULL_ELEMENT_MSG = "MAP comparison not supported for null value elements";
+    private static final int EXPECTED_BYTES_PER_ENTRY = 32;
 
     public MapType(Type keyType, Type valueType)
     {
-        super(parameterizedTypeName("map", keyType.getTypeSignature(), valueType.getTypeSignature()), Slice.class);
+        super(parameterizedTypeName("map", keyType.getTypeSignature(), valueType.getTypeSignature()), Block.class);
         checkArgument(keyType.isComparable(), "key type must be comparable");
         this.keyType = keyType;
         this.valueType = valueType;
     }
 
-    public static Slice toStackRepresentation(Map<?, ?> value, Type keyType, Type valueType)
+    public static Block toStackRepresentation(Map<?, ?> value, Type keyType, Type valueType)
     {
-        BlockBuilder blockBuilder = new VariableWidthBlockBuilder(new BlockBuilderStatus(), 1024);
+        BlockBuilder blockBuilder = new InterleavedBlockBuilder(ImmutableList.of(keyType, valueType), new BlockBuilderStatus(), value.size(), EXPECTED_BYTES_PER_ENTRY);
         for (Map.Entry<?, ?> entry : value.entrySet()) {
             appendToBlockBuilder(keyType, entry.getKey(), blockBuilder);
             appendToBlockBuilder(valueType, entry.getValue(), blockBuilder);
         }
-        return buildStructuralSlice(blockBuilder);
+        return blockBuilder.build();
+    }
+
+    @Override
+    public BlockBuilder createBlockBuilder(BlockBuilderStatus blockBuilderStatus, int expectedEntries, int expectedBytesPerEntry)
+    {
+        return new ArrayBlockBuilder(
+                new InterleavedBlockBuilder(getTypeParameters(), blockBuilderStatus, expectedEntries, expectedBytesPerEntry),
```

Comment:
this should be at least `2 * expectedEntries` probably more since this needs to hold all the entries for all the maps